attachment_delete_restrict/models/ir_attachment.py

```diff
@@ -59,14 +59,14 @@ def _check_custom_delete_attachment(self, model=None, allow_owner_and_admin=Fals
         if allow_owner_and_admin:
             users += self.create_uid
             groups += self.env.ref("base.group_system")
-        allowed_users = groups.users + users
+        allowed_users = groups.users | users
         if self.env.user not in allowed_users:
             return self._raise_delete_attachment_error(allowed_users)
 
     def unlink(self):
         res_models = list(set(self.filtered("res_model").mapped("res_model")))
         if res_models:
-            models = self.env["ir.model"].search([("model", "in", res_models)])
+            models = self.env["ir.model"].sudo().search([("model", "in", res_models)])
```

Comment:
This should be part of the migration, due to the ACL change in 16.0 (or 15.0, I haven't checked).